Improve dock and engage

BrownGenius opened the pull request:
Hi Team,

After Monday's autonomous practice, I dug further in the code and made a few tweaks.  These all need to be tested on the actual bot before pulling into the main branch.

The main issue we noticed is that the **competition** robot tends to oscillate when trying to balance.  I don't _think_ the oscillation is PID related, though, because the **practice** robot balances fine.  What I think is happening is that since the competition bot is top-heavy, when the charge station tilts in the other direction, the front/back of the robot ends up tilting slightly off the charge station.  When it lands back, sometimes, the robot is no longer square with the charge station, which seems to make it much harder to find a balance point.

This branch tries to maintain a fixed angle/yaw while performing the entire dock/engage sequence.  The way it works is:

- When the autonomous routine is first created, the starting angle/yaw of the robot is saved
- The starting angle/yaw is then passed to all of the auto routines that run "Dock and Engage"
- The auto routines then try to maintain the specified angle/yaw during the entire "Dock and Engage" sequence
- For "Mobility, Dock and Engage", the starting angle/yaw is rotated 180 degrees because the robot is approaching the charging station from the _opposite_ direction of where is first starts.

Please take a look at the changes to see if there are any glaring bugs.  We will test this tonight.

BrownGenius commented on src/main/java/bhs/devilbotz/RobotContainer.java:
In this file, I'm reading the startAngle when the autonomous command is requested and then passing it to all routines that use 'DockAndEngage'.

BrownGenius commented on src/main/java/bhs/devilbotz/commands/auto/DockAndEngage.java:
Here, I just added a new DockAndEngage constructor that accepts a targetAngle.  The old routine/constructor works as it did before.

BrownGenius commented on src/main/java/bhs/devilbotz/commands/auto/MobilityDockAndEngage.java:
For MobilityDockAndEngage, since we approach the ramp from the opposite direction of when we first start, we set the targetAngle to be 180 degrees from the startAngle.

BrownGenius commented on src/main/java/bhs/devilbotz/commands/auto/ScoreAndMobility.java:
During drive practice, the driver sometimes forgot to set the distance to a _negative_ value, causing the robot to drive into the scoring grid.  In ScoreAndMobility, the robot is always facing the scoring grid and pushed up against the lanes, so, here, we **always** treat the distance as negative.

We also explicitly stop the drivetrain.

BrownGenius commented on src/main/java/bhs/devilbotz/commands/auto/ScoreDockAndEngage.java:
Similar to ScoreAndMobility above.

_However, we should probably use the same `-(Math.abs(distance))` logic instead of hard-coding -2._

loafdog replied:
agree on fixing hard-coding.
Same comments as in ScoreAndMoblity here.

BrownGenius commented on src/main/java/bhs/devilbotz/commands/drivetrain/BalancePID.java:
Here, I added an optional targetAngle.  The new straightPid is used only if the targetAngle has been set. Otherwise, doesn't explicitly rotate the robot.

loafdog commented on src/main/java/bhs/devilbotz/commands/auto/ScoreAndMobility.java:
```diff
@@ -31,8 +31,11 @@ public ScoreAndMobility(Arm arm, DriveTrain drivetrain, double distance, Gripper
     addCommands(CommandDebug.start());
     addCommands(new AutoScore(arm, drivetrain, gripper));
     addCommands(new DriveStraightPID(drivetrain, -DriveConstants.POSITION_DRIVE_FROM_PORTAL));
+    addCommands(drivetrain.stop());
```
This isn't a command?  How does this work?
Also might help to add some comments explaining what each step/cmd does.. and why.  There is no explanation what the starting pos of robot is.  What does AutoScore do? Why do we drive straight after? And then stop, etc.. Give a plain english explanation of what is supposed to happen in this auto routine.  Include start position, what happens, and end position.

BrownGenius replied:
> This isn't a command? How does this work?
Hi @loafdog, `drivetrain.stop()` is indeed a command (`public Command stop()`), however, it does make sense to rename the method to `stopCommand()`. I'll make that change

> Also might help to add some comments explaining what each step/cmd does.. and why. 
There's already javadoc comments at the top of the file explaining what the command does.  There are javadocs for all sub commands, too....e.g. `AutoScore` also has comments at the top of the class.  Should I move the comments to be above the constructor so that it is displayed as a hover message in vscode?

> There is no explanation what the starting pos of robot is.
This class doesn't have a starting pos.  I believe the methods that do have a startAngle, it is documented.  Let me know if I missed a location where it isn't documented.

loafdog commented on src/main/java/bhs/devilbotz/commands/auto/MobilityDockAndEngage.java:
```diff
@@ -56,7 +61,8 @@ public MobilityDockAndEngage(
     addCommands(drivetrain.followTrajectoryCommand(path, true, false));
-    addCommands(new DockAndEngage(drivetrain, 2));
+    // Since the trajectory causes the robot to turn 180 degrees, we set the targetAngle accordingly
+    addCommands(new DockAndEngage(drivetrain, 2, 180 - startAngle));
```
I'm not great at math... this does not seem intuitively right.  If robot started at 90.. then opposite would be 270? This line would make it still be 90? Can you explain why `180-startAngle` works, in a comment in the code please?

BrownGenius replied:
The `startAngle` is the yaw/angle that the robot was in when the autonomous command is started (i.e. at the beginning of the match).  Since Mobility drives in a curved path, the DockAndEngage should be exactly opposite of the starting orientation.

![image](https://user-images.githubusercontent.com/49494444/225413103-e35e2c3a-2b85-4ed4-8977-3cc10d731bd6.png)

BrownGenius replied:
It's fixed in the latest code.  It just happened to work "as-is" because we were starting at either "0" or "180".  The "90" degree example exposed the obvious logic error.  Tnx!

loafdog commented on src/main/java/bhs/devilbotz/commands/assist/AutoScore.java:
```diff
@@ -38,10 +38,12 @@ public AutoScore(Arm arm, DriveTrain drivetrain, Gripper gripper) {
     addCommands(new ArmMoveDistance(arm, ArmConstants.POSITION_SCORING_DELTA, gripper));
     addCommands(new GripperOpen(gripper));
-    addCommands(CommandDebug.start());
+    addCommands(CommandDebug.end());
```
I see this typo in a few other commands.  Do you want to add those here too? Or fix in sep PR to keep this one smaller?

BrownGenius replied:
Thanks. I'll look for this error and fix in all the other places, too.

BrownGenius replied:
I couldn't find this error anywhere else.

loafdog replied:
sorry, must of been looking at wrong branch/code or out of date code

loafdog reviewed (commented):
The thing about comments for what commands do is that they are at the top of the file, and I have to bounce between the lines of code that do stuff and the comment.   I think javadocs are great if intent is to document APIs or code to be used by other people/things.  But for commenting pieces of code inside a project, I don't think it works well.  This is just my 2 cents, and based on how I work with and read code.

The comments do not say where the robot is expected to be positioned when starting.  Imagine being a brand new person(or someone who missed a few meetings :) ) looking at this and trying to figure out how to test/use these auto routines.   I would not know where to place the robot to start.  What direction should the robot be facing? What piece, if any should be in the claw? What position should the arm and claw be in to start?  I'm being picky here but this is the kind of info needed in comments to make them useful imho.

PR looks good. hopefully testing goes well.

ParkerMeyers commented on src/main/java/bhs/devilbotz/RobotContainer.java:
```diff
@@ -218,20 +219,21 @@ public Command getAutonomousCommand(AutonomousModes autoMode) {
-          autonomousCommand.addCommands(new DockAndEngage(driveTrain, maxDistance));
+          autonomousCommand.addCommands(new DockAndEngage(driveTrain, maxDistance, startAngle));
           break;
         case MOBILITY_DOCK_AND_ENGAGE_HUMAN_SIDE:
           autonomousCommand.addCommands(
               new MobilityDockAndEngage(
-                  driveTrain, CommunityLocation.HUMAN, DriverStation.getAlliance()));
+                  driveTrain, CommunityLocation.HUMAN, DriverStation.getAlliance(), startAngle));
           break;
         case MOBILITY_DOCK_AND_ENGAGE_WALL_SIDE:
           autonomousCommand.addCommands(
               new MobilityDockAndEngage(
-                  driveTrain, CommunityLocation.WALL, DriverStation.getAlliance()));
+                  driveTrain, CommunityLocation.WALL, DriverStation.getAlliance(), startAngle));
           break;
         case SCORE_DOCK_AND_ENGAGE:
-          autonomousCommand.addCommands(new ScoreDockAndEngage(arm, driveTrain, gripper));
+          autonomousCommand.addCommands(
+              new ScoreDockAndEngage(arm, driveTrain, gripper, startAngle));
           break;
         case SCORE_MOBILITY_DOCK_ENGAGE:
```
It would be really nice if we could separate these into separate files, its starting to get quite confusing.

BrownGenius replied:
I agree. I'll move the autonomous routine logic to a different class.

BrownGenius replied:
Done.

ParkerMeyers commented on src/main/java/bhs/devilbotz/commands/assist/PickupFromGround.java:
```diff
@@ -34,7 +34,8 @@ public PickupFromGround(Arm arm, Gripper gripper, DriveTrain drivetrain) {
             // Drive backward
-            new SequentialCommandGroup(new DriveStraightPID(drivetrain, -.3), drivetrain.stop()),
+            new SequentialCommandGroup(
+                new DriveStraightPID(drivetrain, -.3), drivetrain.stopCommand()),
```
```suggestion
                new DriveStraightPID(drivetrain, -0.3), drivetrain.stopCommand()),
```
Just for readability

ParkerMeyers reviewed (approved):
Looks good, just readability changes

BrownGenius commented:
> The thing about comments for what commands do is that they are at the top of the file, and I have to bounce between the lines of code that do stuff and the comment. I think javadocs are great if intent is to document APIs or code to be used by other people/things. But for commenting pieces of code inside a project, I don't think it works well. This is just my 2 cents, and based on how I work with and read code.

Hi @loafdog, I agree.  I'll try to move the comments from above the class to the method/constructor so that hovering over the method/constructor in the SequentalCommand group will actually show relevant information.  I need to see if there's a way to copy javadoc for overloaded constructors.  E.g. I'd hate to have to copy/paste 95% of the same text for each constructor just to describe what the additional parameter does.

> The comments do not say where the robot is expected to be positioned when starting. Imagine being a brand new person(or someone who missed a few meetings :) ) looking at this and trying to figure out how to test/use these auto routines. I would not know where to place the robot to start. What direction should the robot be facing? What piece, if any should be in the claw? What position should the arm and claw be in to start? I'm being picky here but this is the kind of info needed in comments to make them useful imho.
> 
> PR looks good. hopefully testing goes well.

I agree. I will clearly document the expected starting conditions for each autonomous routine.  Unfortunately, testing today was (again) limited due to mechanical/electrical issues.  Dock & Engage is just too unstable for the competition bot.  We may need to switch to just Dock and skip the engage portion.

loafdog reviewed (dismissed):
lgtm.  if this latest set of changes have been tested on robot double lgtm.

BrownGenius commented:
> lgtm. if this latest set of changes have been tested on robot double lgtm.

Thanks. I'm going to hold off on pushing it to main until we've had a chance to test this on Friday in Reading on an actual ramp.  Worst case, I may split this PR into a few smaller ones for the changes that were vetted fully on the robot.

ParkerMeyers reviewed (dismissed):
Much better. Looks good to me!

BrownGenius commented:
> lgtm. if this latest set of changes have been tested on robot double lgtm.

I added a bunch of documentation and addressed a few comments.  I won't push to main until we have a final test Friday evening @ Reading.

ParkerMeyers reviewed (approved):
lgtm
